Capability tracking

umar456 opened the pull request:
This PR adds capability registration and querying through the ValidationState_t class. 

* 5657 unit tests
* Capability checks for OpDecorate, OpVariable and OpTypeMatrix

Additional instructions will be added in future PRs

dekimir commented on source/validate_instruction.cpp:
```diff
@@ -51,11 +324,252 @@ spv_result_t InstructionPass(ValidationState_t& _,
+      case SpvOpGroupCommitWritePipe: break;
       default:
-        break;
+        return _.diag(SPV_ERROR_INTERNAL) << "NOT IMPLEMENTED";
```
Quite apart from the trivial-case explosion above, validation should not fail because the validator is deficient.  That makes it impossible to use the validator in test suites and live drivers.

dekimir commented on source/validate_types.cpp:
```diff
@@ -295,6 +304,86 @@ bool ValidationState_t::in_block() const {
+
+  // Set Base capability
+  module_capabilities_[capability] = true;
+  capability = getDependantCapability(capability);
```
This line looks like dead code to me.

dekimir commented on test/Validate.Capability.cpp:
```diff
@@ -0,0 +1,428 @@
+                          SPV_VALIDATE_INSTRUCTION_BIT>;
+
+TEST_F(ValidateCapability, Default) {
+  char str[] = R"(
```
Nit: `const`.

dekimir commented on test/Validate.Capability.cpp:
```diff
@@ -0,0 +1,428 @@
+  auto& valid_capabilities = operation.second;
+
+  auto it = find(begin(valid_capabilities), end(valid_capabilities), capability);
+  if(it != end(operation.second)) {
```
Nit: I'd use `valid_capabilities` here instead of `operation.second`, for clarity.

dekimir commented:
Looks good, other than the "NOT IMPLEMENTED" abort.

umar456 commented:
Addressed Feedback. Added additional checks and unit tests (total count 12,308 tests).

I am done with this PR for now. I am going to jump to another part of the project.

dekimir commented:
Looks good, just need to fix the Triangles execution mode.  If you're busy with other parts, I can do it during this PR.

I can also fix the ViewportIndex capability in the parsing table.

dekimir commented:
BTW, nice job of adding tens of thousands of new tests!

umar456 commented:
Sure that would be great!

umar456 commented:
Rebased and squashed
